File: src/lib/solver/exact/bounds.ts
import type { ItemId, Recipe } from "@/data/types";
import { legalUnderclocks, recipeCyclesPerMinuteExact } from "./underclocks";
import type { ExactRecipeGraph } from "./recipe-graph";
import { Rational, type RationalInput } from "./rational";

const ZERO = new Rational(BigInt(0));

export type ExactRawAvailability = Partial<Record<ItemId, RationalInput>>;

export interface ExactRecipeBound {
  readonly recipe: Recipe;
  /** Scarce raw items consumed by one cycle of this recipe and its dependencies. */
  readonly rawPerCycle: ReadonlyMap<ItemId, Rational>;
  /** Scarce raw items consumed by one machine at 100% clock for one minute. */
  readonly rawPerEffectiveMachine: ReadonlyMap<ItemId, Rational>;
  readonly maxCyclesPerMinute: Rational;
  readonly maxEffectiveMachines: Rational;
  readonly minimumLegalClock: Rational;
  /**
   * Maximum physical machine count if every active machine uses the recipe's
   * smallest legal clock.
   */
  readonly maxMachines: bigint;
}

function addRate(target: Map<ItemId, Rational>, itemId: ItemId, amount: Rational): void {
  target.set(itemId, (target.get(itemId) ?? ZERO).add(amount));
}

function expandItemToScarceRaws(
  graph: ExactRecipeGraph,
  itemId: ItemId,
  amount: Rational,
  target: Map<ItemId, Rational>,
): void {
  const item = graph.itemById.get(itemId);
  if (!item) {
    throw new Error(`Unknown item in validated recipe graph: ${itemId}`);
  }
  if (item.isRaw) {
    if (!item.isUnlimited) addRate(target, itemId, amount);
    return;
  }

  const producer = graph.producerByItem.get(itemId);
  const output = producer?.outputs[0];
  if (!producer || !output) {
    throw new Error(`No primary producer in validated recipe graph for ${itemId}`);
  }

  const cycles = amount.divide(output.amount);
  for (const input of producer.inputs) {
    expandItemToScarceRaws(graph, input.item, Rational.from(input.amount).multiply(cycles), target);
  }
}

export function rawRequirementsPerRecipeCycle(
  graph: ExactRecipeGraph,
  recipe: Recipe,
): ReadonlyMap<ItemId, Rational> {
  const requirements = new Map<ItemId, Rational>();
  for (const input of recipe.inputs) {
    expandItemToScarceRaws(graph, input.item, Rational.from(input.amount), requirements);
  }
  return requirements;
}

/**
 * Computes finite per-recipe activity and machine bounds from scarce raw
 * availability. Missing scarce resources have zero availability.
 */
export function computeRecipeBounds(
  graph: ExactRecipeGraph,
  rawAvailability: ExactRawAvailability,
): ReadonlyMap<string, ExactRecipeBound> {
  const available = new Map<ItemId, Rational>();
  for (const itemId of graph.scarceRawIds) {
    const amount = Rational.from(rawAvailability[itemId] ?? 0);
    if (amount.compare(0) < 0) {
      throw new RangeError(`Raw availability cannot be negative: ${itemId}`);
    }
    available.set(itemId, amount);
  }

  const bounds = new Map<string, ExactRecipeBound>();
  for (const recipe of graph.topologicalRecipes) {
    const rawPerCycle = rawRequirementsPerRecipeCycle(graph, recipe);
    let maxCyclesPerMinute: Rational | null = null;
    for (const [itemId, requirement] of rawPerCycle) {
      if (requirement.compare(0) <= 0) continue;
      const candidate = (available.get(itemId) ?? ZERO).divide(requirement);
      if (maxCyclesPerMinute === null || candidate.compare(maxCyclesPerMinute) < 0) {
        maxCyclesPerMinute = candidate;
      }
    }
    if (maxCyclesPerMinute === null) {
      throw new Error(`Recipe ${recipe.id} has no scarce raw requirement`);
    }

    const cyclesPerMinute = recipeCyclesPerMinuteExact(recipe);
    const maxEffectiveMachines = maxCyclesPerMinute.divide(cyclesPerMinute);
    const rawPerEffectiveMachine = new Map<ItemId, Rational>();
    for (const [itemId, requirement] of rawPerCycle) {
      rawPerEffectiveMachine.set(itemId, requirement.multiply(cyclesPerMinute));
    }

    const clocks = legalUnderclocks(recipe);
    const minimumLegalClock = clocks[0];
    if (!minimumLegalClock) {
      throw new Error(`Recipe ${recipe.id} has no legal underclock`);
    }

    bounds.set(recipe.id, {
      recipe,
      rawPerCycle,
      rawPerEffectiveMachine,
      maxCyclesPerMinute,
      maxEffectiveMachines,
      minimumLegalClock,
      maxMachines: maxEffectiveMachines.divide(minimumLegalClock).floor(),
    });
  }
  return bounds;
}

File: rust/solver/src/bounds.rs
//! Raw-derived per-recipe activity bounds, ported from
//! `src/lib/solver/exact/bounds.ts`.

use std::collections::BTreeMap;

use num_bigint::BigInt;
use num_rational::BigRational;
use num_traits::Zero;

use crate::graph::RecipeGraph;
use crate::rational::floor_bigint;
use crate::underclocks::{legal_underclocks, recipe_cycles_per_minute};

#[derive(Debug, Clone)]
pub struct RecipeBound {
    pub recipe: usize,
    /// Scarce raw items consumed by one cycle of this recipe and its dependencies.
    pub raw_per_cycle: BTreeMap<usize, BigRational>,
    pub max_cycles_per_minute: BigRational,
    pub max_effective_machines: BigRational,
    pub minimum_legal_clock: BigRational,
    pub max_machines: BigInt,
}

fn expand_item_to_scarce_raws(
    graph: &RecipeGraph,
    item: usize,
    amount: BigRational,
    target: &mut BTreeMap<usize, BigRational>,
) -> Result<(), String> {
    let item_data = &graph.items[item];
    if item_data.is_raw {
        if !item_data.is_unlimited {
            let entry = target.entry(item).or_insert_with(BigRational::zero);
            *entry += amount;
        }
        return Ok(());
    }

    let producer = graph
        .producer_by_item
        .get(&item)
        .ok_or_else(|| format!("No primary producer for {}", item_data.id))?;
    let producer_recipe = &graph.recipes[*producer];
    let output = producer_recipe
        .outputs
        .first()
        .ok_or_else(|| format!("Producer {} has no output", producer_recipe.id))?;

    let cycles = amount / output.amount.clone();
    for input in &producer_recipe.inputs {
        expand_item_to_scarce_raws(graph, input.item, input.amount.clone() * cycles.clone(), target)?;
    }
    Ok(())
}

pub fn raw_requirements_per_recipe_cycle(
    graph: &RecipeGraph,
    recipe: usize,
) -> Result<BTreeMap<usize, BigRational>, String> {
    let mut requirements = BTreeMap::new();
    for input in &graph.recipes[recipe].inputs {
        expand_item_to_scarce_raws(graph, input.item, input.amount.clone(), &mut requirements)?;
    }
    Ok(requirements)
}

/// Computes finite per-recipe activity and machine bounds from scarce raw
/// availability. Missing scarce resources have zero availability.
pub fn compute_recipe_bounds(
    graph: &RecipeGraph,
    raw_availability: &BTreeMap<usize, BigRational>,
) -> Result<BTreeMap<usize, RecipeBound>, String> {
    let mut bounds = BTreeMap::new();
    for &recipe_index in &graph.topological_recipes {
        let recipe = &graph.recipes[recipe_index];
        let raw_per_cycle = raw_requirements_per_recipe_cycle(graph, recipe_index)?;

        let mut max_cycles_per_minute: Option<BigRational> = None;
        for (item, requirement) in &raw_per_cycle {
            if *requirement <= BigRational::zero() {
                continue;
            }
            let available = raw_availability.get(item).cloned().unwrap_or_else(BigRational::zero);
            let candidate = available / requirement.clone();
            if max_cycles_per_minute
                .as_ref()
                .map(|current| candidate < *current)
                .unwrap_or(true)
            {
                max_cycles_per_minute = Some(candidate);
            }
        }
        let max_cycles_per_minute = max_cycles_per_minute
            .ok_or_else(|| format!("Recipe {} has no scarce raw requirement", recipe.id))?;

        let cycles_per_minute = recipe_cycles_per_minute(recipe);
        let max_effective_machines = max_cycles_per_minute.clone() / cycles_per_minute;

        let clocks = legal_underclocks(recipe);
        let minimum_legal_clock = clocks
            .first()
            .cloned()
            .ok_or_else(|| format!("Recipe {} has no legal underclock", recipe.id))?;

        let max_machines =
            floor_bigint(&(max_effective_machines.clone() / minimum_legal_clock.clone()));
        let _ = bounds.insert(
            recipe_index,
            RecipeBound {
                recipe: recipe_index,
                raw_per_cycle,
                max_cycles_per_minute,
                max_effective_machines,
                minimum_legal_clock,
                max_machines,
            },
        );
    }
    Ok(bounds)
}
